url74b2be4e76_gurbano_petulant_octo_sudoku_tgz-pJ8-test_testJ8/src/ocr/MainEntry.java

```diff
@@ -0,0 +1,639 @@
+    *
+    * @param event The event
+    */
+   void add_actionPerformed(java.awt.event.ActionEvent event)
+   {
+     int i;
+
+     String letter = JOptionPane.showInputDialog(
+       "Please enter a letter you would like to assign this sample to.");
+     if ( letter==null )
+       return;
+
+     if ( letter.length()>1 ) {
+       JOptionPane.showMessageDialog(this,
+                                     "Please enter only a single letter.","Error",
+                                     JOptionPane.ERROR_MESSAGE);
+       return;
+     }
+
+     entry.downSample();
+     SampleData sampleData = (SampleData)sample.getData().clone();
+     sampleData.setLetter(letter.charAt(0));
+
+     for ( i=0;i<letterListModel.size();i++ ) {
+       Comparable str = (Comparable)letterListModel.getElementAt(i);
+       if ( str.equals(letter) ) {
+         JOptionPane.showMessageDialog(this,
+                                       "That letter is already defined, delete it first!","Error",
+                                       JOptionPane.ERROR_MESSAGE);
+         return;
+       }
+
+       if ( str.compareTo(sampleData)>0 ) {
+         letterListModel.add(i,sampleData);
+         return;
+       }
+     }
+     letterListModel.add(letterListModel.size(),sampleData);
+     letters.setSelectedIndex(i);
+     entry.clear();
+     sample.repaint();
+
+   }
+
+   /**
+    * Called when the del button is pressed.
+    *
+    * @param event The event.
+    */
+   void del_actionPerformed(java.awt.event.ActionEvent event)
+   {
+     int i = letters.getSelectedIndex();
+
+     if ( i==-1 ) {
+       JOptionPane.showMessageDialog(this,
+                                     "Please select a letter to delete.","Error",
+                                     JOptionPane.ERROR_MESSAGE);
+       return;
+     }
+
+     letterListModel.remove(i);
+   }
+
+   class SymListSelection implements javax.swing.event.ListSelectionListener {
+     public void valueChanged(javax.swing.event.ListSelectionEvent event)
+     {
+       Object object = event.getSource();
+       if ( object == letters )
+         letters_valueChanged(event);
+     }
+   }
+
+   /**
+    * Called when a letter is selected from the list box.
+    *
+    * @param event The event
+    */
+   void letters_valueChanged(javax.swing.event.ListSelectionEvent event)
+   {
+     if ( letters.getSelectedIndex()==-1 )
+       return;
+     SampleData selected =
+       (SampleData)letterListModel.getElementAt(letters.getSelectedIndex());
+     sample.setData((SampleData)selected.clone());
+     sample.repaint();
+     entry.clear();
+
+   }
+
+   /**
+    * Called when the load button is pressed.
+    *
+    * @param event The event
+    */
+   void load_actionPerformed(java.awt.event.ActionEvent event)
+   {
+     try {
+       FileReader f;// the actual file stream
+       BufferedReader r;// used to read the file line by line
+
+       f = new FileReader( new File("./sample.dat") );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *f*. The resource is referenced at lines 400, 406. The resource is closed at lines 420, 421. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *f* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).